File: javascript-solutions/leetcode/strStr.js
/// # 28. Implement strStr()
/// Return the index of the first occurrence of `needle` in `haystack`, or -1 if `needle` is not
/// part of `haystack`
///
/// if `needle` is the empty string, return 0.
/// if `needle` is not in haystack, return -1.
///


/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */
const strStr = function(haystack, needle) {
    if (needle.length === 0) return 0;

    let i = 0;
    while (i < haystack.length) {
        if (haystack.charAt(i) === needle.charAt(0) && allCharsMatch(haystack, i, needle)) {
            return i;
        }
        i++;
    }

    return -1;
};

/// returns true if all chars of `subStr` match the chars of `str`, starting at `fromIndex` of `str`, else false.
function allCharsMatch(str, fromIndex, subStr) {
    const sliceLength = str.length - fromIndex;
    if (subStr.length > sliceLength || sliceLength <= 0 ) return false

    let i = fromIndex;
    let j = 0;

    while (i < str.length && j < subStr.length) {

        if (str.charAt(i) !== subStr.charAt(j)) {
            return false;
        }
        i++;
        j++;
    }

    return true;
}

console.log( strStr("hello", "ll") );
console.log( strStr("hello", "llz") );
console.log( strStr("hello", "") );
console.log( strStr("hello", "hello") );

File: rust-solutions/leetcode/src/bin/str-str.rs
/// # 28. Implement strStr()
/// Return the index of the first occurrence of `needle` in `haystack`, or -1 if `needle` is not
/// part of `haystack`
///
/// if `needle` is the empty string, return 0.
/// if `needle` is not in haystack, return -1.
///

pub fn str_str(haystack: String, needle: String) -> i32 {
    if needle.is_empty() {
        return 0;
    }

    // assuming haystack and needle contain ASCII chars only
    for i in 0..haystack.len() {
        if i + needle.len() <= haystack.len() && haystack[i..(i + needle.len())] == needle[..] {
            return i as i32;
        }
    }

    -1
}

fn main() {}

#[cfg(test)]
mod tests {
    use super::str_str;

    #[test]
    fn has_ll() {
        assert_eq!(str_str(String::from("hello"), String::from("ll")), 2);
    }

    #[test]
    fn has_lo_at_end() {
        assert_eq!(str_str(String::from("hello"), String::from("lo")), 3);
    }

    #[test]
    fn has_equal_strings() {
        assert_eq!(str_str(String::from("hello"), String::from("hello")), 0);
    }

    #[test]
    fn needle_is_empty() {
        assert_eq!(str_str(String::from("hello"), String::from("")), 0);
    }

    #[test]
    fn needle_not_in_haystack() {
        assert_eq!(str_str(String::from("hello"), String::from("zip")), -1);
    }
}
